clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,449 @@
+		dataElem.Data[perc95] = percentile.percentile95
+		klog.Info("Aggregate Metric value: ", aggregatePodPairMetrics)
+	}
+	klog.Infof("PodPairNumber: %v, Response received number: %v ", len(npm.podInfo.uniquePodPairList), len(npm.metricVal))
+}
+
+func (npm *networkPerformanceMeasurement) createResultSummary() {
+	var resultSummary testResultSummary
+	switch npm.protocol {
+	case protocolTCP:
+		npm.getMetricData(&resultSummary, tcpBandwidth, throughput)
+		resultSummary.protocol = protocolTCP
+	case protocolUDP:
+		npm.getMetricData(&resultSummary, udpPacketPerSecond, packetPerSecond)
+		npm.getMetricData(&resultSummary, udpJitter, jitter)
+		npm.getMetricData(&resultSummary, udpLatencyAverage, latency)
+		resultSummary.protocol = protocolUDP
+	case protocolHTTP:
+		npm.getMetricData(&resultSummary, httpResponseTime, responseTime)
+		resultSummary.protocol = protocolHTTP
+	}
+	resultSummary.service = "P2P"
+	resultSummary.podRatio = npm.podRatioType
+	npm.resultSummary = resultSummary
+}
+
+func (npm *networkPerformanceMeasurement) getMetricData(data *testResultSummary, metricIndex int, metricName string) {
+	var metricDataItem measurementutil.DataItem
+	metricDataItem.Data = make(map[string]float64)
+	metricDataItem.Labels = make(map[string]string)
+	metricDataItem.Labels["Metric"] = metricName
+	npm.calculateMetricDataValue(&metricDataItem, metricIndex)
+	metricDataItem.Unit = metricUnitMap[metricDataItem.Labels["Metric"]]
+	data.DataItems = append(data.DataItems, metricDataItem)
+	klog.V(3).Infof("TestResultSummary: %v", data)
+}
+
+func (wp *workerPodInfo) formUniquePodPair() {
+	nh := make(nodeHeap, len(wp.workerPodMap))
+	i := 0
+	for key, value := range wp.workerPodMap {
+		nh[i] = &workerNodeInfo{nodeName: key, numberOfPods: len(value)}
+		i++
+	}
+	heap.Init(&nh)
+	for i := 0; nh.Len() > 0; i++ {
+		firstNode := heap.Pop(&nh).(*workerNodeInfo)
+		if firstNode.numberOfPods > 1 {
+			heap.Push(&nh, &workerNodeInfo{nodeName: firstNode.nodeName, numberOfPods: firstNode.numberOfPods - 1})
+		}
+		secondNode := heap.Pop(&nh).(*workerNodeInfo)
+		if secondNode.numberOfPods > 1 {
+			heap.Push(&nh, &workerNodeInfo{nodeName: secondNode.nodeName, numberOfPods: secondNode.numberOfPods - 1})
+		}
+		if err := wp.populateUniquePodPairList(i, firstNode, secondNode); err != nil {
+			klog.Error(err)
+		}
+	}
+}
+
+func (wp *workerPodInfo) populateUniquePodPairList(index int, firstNode *workerNodeInfo, secondNode *workerNodeInfo) error {
+	sourcePod, err := wp.Pop(firstNode.nodeName)
+	if err != nil {
+		return err
+	}
+	destinationPod, err := wp.Pop(secondNode.nodeName)
+	if err != nil {
+		return err
+	}
+	uniquePodPair := uniquePodPair{
+		pairIndex:          index,
+		sourcePodName:      sourcePod.podName,
+		sourcePodIP:        sourcePod.podIP,
+		destinationPodName: destinationPod.podName,
+		destinationPodIP:   destinationPod.podIP}
+	wp.uniquePodPairList = append(wp.uniquePodPairList, uniquePodPair)
```

Comment:
This shouldn't be stored in the type field.

We should simply return the list directly from the `formUniquePodPair` function, i.e. the signature of formUniquePodPairs should be:

```
(wp *workerPodInfo) formUniquePodPair() []uniquePodPair
```

Reply:
modified